File: atcoder/abs/09-daydream/ruby/main.rb
# Copyright (c) 2023 Furukawa, Atsushi <atsushifx@aglabo.com>
#
# This software is released under the MIT License.
# https://opensource.org/licenses/MIT

## initialize
Keywords = [
  "dream",
  "dreamer",
  "erase",
  "eraser",
]

## solve


## solver function
def isMatchKeywords(str)
  ret = false
  if (str == "") then
    ret = true
  else
    ret = false
    Keywords.each do |k|
      idx = str.index(k)
      if (idx === 0) then
        ret2 = isMatchKeywords(str.delete_prefix(k))
        if (ret2) then
          ret = true
          break
        else

        end
      end
    end
  end
  return ret
end

## main
s = gets.chomp
print (isMatchKeywords(s) ? "YES" : "NO"),"\n"

File: atcoder/abs/09-daydream/go/main.go
// Copyright (c) 2023 Furukawa, Atsushi <atsushifx@aglabo.com>
//
// This software is released under the MIT License.
// https://opensource.org/licenses/MIT

package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

// Global Variables
const IOBUFFSIZE = 10 ^ 6

var Stdin = bufio.NewScanner(os.Stdin)
var Stdout = bufio.NewWriter(os.Stdout)

// atcoder lib
func getNext() string {
	Stdin.Scan()
	return strings.TrimSpace(Stdin.Text())
}

func getNextInt() int64 {
	var i int64
	var e error
	i, e = strconv.ParseInt(getNext(), 10, 64)
	if e != nil {
		panic(e)
	}
	return i
}

// initialize
func initialize() {
	stdinBuffer := make([]byte, IOBUFFSIZE)
	Stdin.Buffer(stdinBuffer, IOBUFFSIZE)
	Stdin.Split(bufio.ScanWords)
}

// main
func main() {
	initialize()
	defer Stdout.Flush()
	solve()
	Stdout.Flush()
}

// problem solver
func solve() {
	// initialize
	var Keywords = []string{
		"dream",
		"dreamer",
		"erase",
		"eraser",
	}

	// input
	var S string = getNext()

	// solve
	var match bool = isMatch(S, Keywords)

	// output
	if match {
		fmt.Println("YES")
	} else {
		fmt.Println("NO")
	}

}

// solver function
func isMatch(str string, keywords []string) bool {
	var ret bool = false
	for _, key := range keywords {
		if str == key {
			ret = true
			break
		} else if strings.HasSuffix(str, key) {
			var str2 string = fmt.Sprint(str[:len(str)-len(key)])
			ret = isMatch(str2, keywords)
			break
		} else {

		}
	}
	return ret
}
